integration-test/src/test/java/org/apache/iotdb/db/it/utils/TestUtils.java

```diff
@@ -0,0 +1,250 @@
+        ResultSetMetaData resultSetMetaData = resultSet.getMetaData();
+        for (int i = 1; i <= resultSetMetaData.getColumnCount(); i++) {
+          assertEquals(expectedHeader[i - 1], resultSetMetaData.getColumnName(i));
+        }
+
+        cnt = 0;
+        while (resultSet.next()) {
+          StringBuilder builder = new StringBuilder();
+          for (int i = 1; i <= resultSetMetaData.getColumnCount(); i++) {
+            builder.append(resultSet.getString(expectedHeader[i - 1])).append(",");
+          }
+          assertEquals(expectedRetArray[cnt], builder.toString());
+          cnt++;
+        }
+        Assert.assertEquals(expectedRetArray.length, cnt);
+      }
+
+      try (ResultSet resultSet = statement.executeQuery(sql + " order by time desc")) {
+        ResultSetMetaData resultSetMetaData = resultSet.getMetaData();
+        for (int i = 1; i <= resultSetMetaData.getColumnCount(); i++) {
+          assertEquals(expectedHeader[i - 1], resultSetMetaData.getColumnName(i));
+        }
+
+        cnt = expectedRetArray.length;
+        while (resultSet.next()) {
+          StringBuilder builder = new StringBuilder();
+          for (int i = 1; i <= resultSetMetaData.getColumnCount(); i++) {
+            builder.append(resultSet.getString(expectedHeader[i - 1])).append(",");
+          }
+          assertEquals(expectedRetArray[cnt - 1], builder.toString());
+          cnt--;
+        }
+        Assert.assertEquals(0, cnt);
+      }
+    } catch (Exception e) {
+      e.printStackTrace();
+      fail(e.getMessage());
+    }
+  }
+
+  public static void assertTestFail(String sql, String errMsg) {
+    try (Connection connection = EnvFactory.getEnv().getConnection();
+        Statement statement = connection.createStatement()) {
+      statement.executeQuery(sql);
+      fail();
+    } catch (Exception e) {
+      Assert.assertTrue(e.getMessage(), e.getMessage().contains(errMsg));
+    }
+  }
+
+  public static void assertResultSetEqual(
+      ResultSet actualResultSet, String expectedHeader, String[] expectedRetArray)
+      throws SQLException {
+    assertResultSetEqual(actualResultSet, expectedHeader, expectedRetArray, null);
+  }
+
+  public static void assertResultSetEqual(
+      ResultSet actualResultSet, String expectedHeader, Set<String> expectedRetSet)
+      throws SQLException {
+    ResultSetMetaData resultSetMetaData = actualResultSet.getMetaData();
+    StringBuilder header = new StringBuilder();
+    for (int i = 1; i <= resultSetMetaData.getColumnCount(); i++) {
+      header.append(resultSetMetaData.getColumnName(i)).append(",");
+    }
+    assertEquals(expectedHeader, header.toString());
+
+    while (actualResultSet.next()) {
+      StringBuilder builder = new StringBuilder();
+      for (int i = 1; i <= resultSetMetaData.getColumnCount(); i++) {
+        builder.append(actualResultSet.getString(i)).append(",");
+      }
+      assertTrue(expectedRetSet.contains(builder.toString()));
```

Comment:
As this is an assertion in Set traversing with an unpredictable order, we should put a more detailed message in `assertTrue(String message, boolean condition)` to show which string does it miss.
What's more, how about building an `actualHashSet` storing all the data of the dataset, and use `Assert.equals` to compare the two sets directly? So we don't need to mutate the expectedRetSet any more. The only con is the output may be too long if the dataset is large.

Reply:
fixed